04-cs-fundamentals/classroom/Finite-State-Machines.md

```diff
@@ -0,0 +1,182 @@
+  --- | --- 
+  Farm Sensor - Timer Expired | FS - TEX
+  Farm Sensor - Timer Running | FS - TR
+  No Farm Sensor - Timer Running | NFS - TR
+  No Farm Sensor - Timer Expired | NFS - TEX
+
+**Bonus**
+
+Can you write a FSM which will determine if given the alphabet {a, b} that the input has an even number of the letter `a` or `b`?
+
+## Summary
+
+A Finite State Machine is a mathematical way to model computation.  The model consists of a limited number of states and a finite alphabet of possible input characters.  The machine transitions from state to state as it reads in each item of input.  An FSM does not remember previous input, but only the current state it is in.  
+
+FSMs are used in a variety of applications and can be an effective way to model and understand systems which consist of a limited number of potential states.  
+
+## Further Concepts
+
+An FSM isn't a complete computational system.  Full computers can **remember** more than one particular value.  To model this mathematicains have developed additional systems which build on FSMs.
+
+### Nondeterministic Finite State Machines
+
+A nondeterministic finite state machine is a machine which can take multiple paths given a specific input.  The machine will accept input if there is *at least* one path it can follow to accept the input.  The FSMs we've been using are called Deterministic Finite State Machines as they can only take one path for each given input.  It's been shown that you can build a deterministic finite state machine that is equivalent to a nondeterministic state machine.
+
+### Pushdown Automata
+
+A Pushdown Automata is a finite state machine that also has access to a **stack**.  Because it can access the stack to remember past inputs or states a Pushdown Automata is more powerful than a finite state machine.
+
+### Turing Machines
+
+[Alan Turing](https://en.wikipedia.org/wiki/Alan_Turing) is one of the father's of computer Science and developed this "complete" model of computation.  In a Turing machine the machine has an infinitely long "tape" of cells.  It can move to individual cells, read and write the value at each cell.  With this the machine has infinite storage capcacity.  In real life no machine has infinite storage, but modern computer systems usually have much more than is needed, and a Turing Machine has been shown to be able to compute an answer to any problem a computer can solve (speed is another issue).  Thus the term **Turing Complete** is used to indicate a programming language that can accomplish all tasks that a computer can accomplish, if you ignore the limitations of memory.  
+
+Turing machines were used in mathematical proofs to illustrate a class of problems which are not computable.  In other words, you cannot write a computer program to solve them in all cases.  A classic example is that you cannot write a program which will read in another program and predict with perfect accuracy if that program will ever end (no infinite loops).
```

Comment:
IIRC it's not even possible to probabilistically predict whether or not a machine will halt correctly.  (I asked a professor about this once but it was a while ago so perhaps I'm mis-remembering or the state of research has changed.)

Reply:
You can't predict if it will halt correctly or even if will ever  halt actually.  I had to do the proof on this for my final once.